tutorials/getting_started/laziness.md

```diff
@@ -0,0 +1,233 @@
+
+```python
+    def __init__(self, lazy: bool = False) -> None:
+        self.lazy = lazy
+```
+
+This means that if you want your `DatasetReader` subclass to allow for laziness,
+its constructor needs to pass a `lazy` value to the `DatasetReader` constructor,
+and its `from_params` method needs to allow for such a parameter. All of the
+dataset readers included with AllenNLP are built this way.
+
+For example, look at the
+[`SnliReader`](https://github.com/allenai/allennlp/blob/master/allennlp/data/dataset_readers/snli.py).
+
+Its constructor takes a `lazy` parameter and passes it to the superclass constructor:
+
+```python
+    def __init__(self,
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 lazy: bool = False) -> None:
+        super().__init__(lazy)
+        self._tokenizer = tokenizer or WordTokenizer()
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+```
+
+And its `from_params` looks for a `lazy` configuration parameter
+(defaulting to `False` if it's not specified):
+
+```python
+    @classmethod
+    def from_params(cls, params: Params) -> 'SnliReader':
+        tokenizer = Tokenizer.from_params(params.pop('tokenizer', {}))
+        token_indexers = TokenIndexer.dict_from_params(params.pop('token_indexers', {}))
+        lazy = params.pop('lazy', False)
+        params.assert_empty(cls.__name__)
+        return SnliReader(tokenizer=tokenizer,
+                          token_indexers=token_indexers,
+                          lazy=lazy)
+```
+
+Any dataset reader that you want to handle lazy datasets should behave likewise.
+
+## Laziness in `DatasetReader.read`
+
+The primary public interface of the `DatasetReader` is its `read` method, which is implemented in
+the base class (and which I've stripped down to its essence):
+
+```python
+    def read(self, file_path: str) -> Iterable[Instance]:
+        if self.lazy:
+            return _LazyInstances(lambda: iter(self._read(file_path)))
+        else:
+            return ensure_list(self._read(file_path))
+```
+
+In both cases, it calls the private `_read` method, which you have to implement,
+and which itself returns an `Iterable[Instance]`. More on that below.
+
+If the dataset reader was instantiated with `lazy=False`,
+it just makes sure that the returned instances are in a `list` and returns that list.
+That list is loaded into memory and (because it's a list) can be iterated over repeatedly
+after the initial call to `.read()`.
+
+The more interesting case is when the dataset reader was instantiated with `lazy=True`.
+In that case we return a `_LazyInstances` initialized with a lambda function
+that calls `_read()`. `_LazyInstances` is just a simple class wrapper to produce
+lazy iterables (again, I stripped it down to just its essence):
+
+```python
+class _LazyInstances(Iterable):
```

Comment:
This is pretty detailed and involves a private AllenNLP class. I'm not sure the details really belong here?

Reply:
* This level of detail

Reply:
On the other hand, this is a tutorial, not a how-to, and the goal is to teach the reader about how the library works.  We _definitely_ want to mention in here that you'll get multiple calls to `_read()` with `lazy=True`; I could go either way on whether to include this code to demonstrate that.